File: src-tauri/src/domain.rs
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const MAX_VISIBLE_TURNS: usize = 100;
pub const MAX_TRANSCRIPT_CHARS: usize = 120_000;
pub const MAX_SEMANTIC_UNITS: usize = 300;
pub const DEFAULT_MODEL: &str = "gpt-5-mini";
pub const CODEX_CLI_MODEL: &str = "gpt-5.6-luna";
pub const PROMPT_VERSION: &str = "dialogue-atlas-v2";
pub const SCHEMA_VERSION: &str = "1.0.0";
pub const DIALOGUE_ACTS: &[&str] = &[
    "提问",
    "请求",
    "任务",
    "建议",
    "陈述",
    "回答",
    "解释",
    "论证",
    "举例",
    "评价",
    "同意",
    "质疑",
    "纠正",
    "承诺",
    "反馈",
    "话语管理",
    "约束",
    "假设检验",
    "证据",
    "区分",
    "归类",
    "反例",
    "限定",
    "撤回",
    "其他",
];
pub const RELATION_KINDS: &[&str] = &[
    "回应",
    "支持",
    "理由",
    "举例",
    "条件",
    "对比",
    "质疑",
    "反证",
    "收窄",
    "修正",
    "重新归类",
    "撤回",
    "重新打开",
    "中断后续答",
    "导致",
    "未解决",
];

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Speaker {
    User,
    Assistant,
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TimeStatus {
    Valid,
    #[default]
    Missing,
    Invalid,
}

impl TimeStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Valid => "valid",
            Self::Missing => "missing",
            Self::Invalid => "invalid",
        }
    }
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TimeCoverage {
    Complete,
    Partial,
    #[default]
    None,
}

impl TimeCoverage {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Complete => "complete",
            Self::Partial => "partial",
            Self::None => "none",
        }
    }
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SourceFormat {
    RawRollout,
    VisibleExport,
    Paste,
    #[default]
    LegacyUnknown,
}

impl SourceFormat {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::RawRollout => "raw_rollout",
            Self::VisibleExport => "visible_export",
            Self::Paste => "paste",
            Self::LegacyUnknown => "legacy_unknown",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RedactionRange {
    pub kind: String,
    pub original_start_utf16: usize,
    pub original_end_utf16: usize,
    pub redacted_start_utf16: usize,
    pub redacted_end_utf16: usize,
    pub replacement: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SourceMessage {
    pub id: String,
    pub speaker: Speaker,
    pub phase: Option<String>,
    pub sequence: usize,
    pub external_message_id: Option<String>,
    #[serde(default)]
    pub external_turn_id: Option<String>,
    pub source_event_index: Option<usize>,
    #[serde(default)]
    pub occurred_at_utc: Option<DateTime<Utc>>,
    #[serde(default)]
    pub occurred_at_raw: Option<String>,
    #[serde(default)]
    pub time_status: TimeStatus,
    pub text: String,
    pub text_sha256: String,
    pub redacted_text: String,
    #[serde(default)]
    pub redaction_map: Vec<RedactionRange>,
    /// Preview-only display metadata; reconstructed from visible turns after load.
    #[serde(default)]
    pub turn_ordinal: usize,
    #[serde(default)]
    pub operation_only: bool,
    #[serde(default)]
    pub redactions: Vec<PreviewRedaction>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PreviewRedaction {
    pub start: usize,
    pub end: usize,
    pub replacement: String,
    pub kind: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct VisibleTurn {
    pub id: String,
    pub ordinal: usize,
    pub speaker: Speaker,
    pub operation_only: bool,
    pub message_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PrivacyFinding {
    pub message_id: String,
    pub kind: String,
    pub start_utf16: usize,
    pub end_utf16: usize,
    pub preview: String,
    pub replacement: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportPreview {
    pub id: String,
    pub title: String,
    pub source_kind: String,
    #[serde(default)]
    pub source_format: SourceFormat,
    pub source_path: Option<String>,
    pub source_sha256: String,
    #[serde(default)]
    pub external_session_id: Option<String>,
    #[serde(default)]
    pub first_visible_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_activity_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_completed_turn_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_message_id: Option<String>,
    #[serde(default)]
    pub time_coverage: TimeCoverage,
    #[serde(default)]
    pub source_file_size: Option<u64>,
    #[serde(default)]
    pub source_file_mtime_ns: Option<i64>,
    #[serde(default)]
    pub supersedes_conversation_id: Option<String>,
    #[serde(default)]
    pub source_still_writing: bool,
    pub messages: Vec<SourceMessage>,
    pub turns: Vec<VisibleTurn>,
    pub character_count: usize,
    pub privacy_findings: Vec<PrivacyFinding>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CommitImportRequest {
    pub title: String,
    pub preview: ImportPreview,
    /// When true, analysis uses `redacted_text`; the immutable local source still
    /// keeps the exact visible text for evidence inspection.
    #[serde(default = "default_true")]
    pub analyze_redacted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PreviewMessageConfirmation {
    pub id: String,
    pub speaker: Speaker,
    pub phase: Option<String>,
    pub turn_ordinal: usize,
    pub text: String,
    #[serde(default)]
    pub operation_only: bool,
    #[serde(default)]
    pub redactions: Vec<PreviewRedaction>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CommitImportOptions {
    pub preview_id: String,
    pub title: String,
    pub messages: Vec<PreviewMessageConfirmation>,
    pub redaction_enabled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CommitImportResponse {
    pub conversation_id: String,
    #[serde(default)]
    pub already_imported: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct StartAnalysisOptions {
    pub conversation_id: String,
    pub model_id: String,
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AnalysisProviderKind {
    CodexCli,
    #[default]
    OpenaiApi,
}

impl AnalysisProviderKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::CodexCli => "codex_cli",
            Self::OpenaiApi => "openai_api",
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum PlatformKind {
    Macos,
    Windows,
    Other,
}

impl PlatformKind {
    pub fn label(self) -> &'static str {
        match self {
            Self::Macos => "macOS",
            Self::Windows => "Windows",
            Self::Other => "系统",
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CredentialStoreKind {
    MacosKeychain,
    WindowsCredentialManager,
    SystemKeyring,
}

impl CredentialStoreKind {
    pub fn label(self) -> &'static str {
        match self {
            Self::MacosKeychain => "macOS Keychain",
            Self::WindowsCredentialManager => "Windows 凭据管理器",
            Self::SystemKeyring => "系统凭据库",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PlatformCapabilities {
    pub platform: PlatformKind,
    pub available_providers: Vec<AnalysisProviderKind>,
    pub credential_store: CredentialStoreKind,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisSettings {
    pub provider: AnalysisProviderKind,
    pub default_openai_model: String,
    pub codex_cli_model: String,
    pub capabilities: PlatformCapabilities,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisProviderStatus {
    pub provider: AnalysisProviderKind,
    pub ok: bool,
    pub configured: bool,
    pub available: bool,
    pub authenticated: bool,
    pub model: String,
    pub version: Option<String>,
    pub message: String,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ConversationSummary {
    pub id: String,
    pub title: String,
    pub source_kind: String,
    #[serde(default)]
    pub source_format: SourceFormat,
    #[serde(default)]
    pub external_session_id: Option<String>,
    pub turn_count: usize,
    pub character_count: usize,
    pub analyze_redacted: bool,
    #[serde(default)]
    pub first_visible_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_activity_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_completed_turn_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub last_message_id: Option<String>,
    #[serde(default)]
    pub time_coverage: TimeCoverage,
    #[serde(default)]
    pub source_file_size: Option<u64>,
    #[serde(default)]
    pub source_file_mtime_ns: Option<i64>,
    #[serde(default)]
    pub supersedes_conversation_id: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CompletionState {
    Completed,
    InProgressOrUnknown,
    Undated,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CalendarSourceState {
    Active,
    Archived,
    Missing,
    ImportOnly,
}

impl CalendarSourceState {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Active => "active",
            Self::Archived => "archived",
            Self::Missing => "missing",
            Self::ImportOnly => "import_only",
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CalendarImportState {
    NotImported,
    ImportedCurrent,
    SourceUpdated,
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CalendarAnalysisState {
    #[default]
    None,
    Ready,
    Partial,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CalendarEntry {
    pub id: String,
    pub external_session_id: Option<String>,
    pub title: String,
    pub last_activity_at: Option<DateTime<Utc>>,
    pub last_completed_turn_at: Option<DateTime<Utc>>,
    pub completion_state: CompletionState,
    pub source_state: CalendarSourceState,
    pub import_state: CalendarImportState,
    pub analysis_state: CalendarAnalysisState,
    pub imported_version_count: usize,
    pub snapshot_count: usize,
    pub latest_conversation_id: Option<String>,
    #[serde(default)]
    pub turn_count: Option<usize>,
    #[serde(default)]
    pub active_day_count: Option<usize>,
    #[serde(default)]
    pub time_coverage: Option<TimeCoverage>,
    #[serde(default)]
    pub source_path: Option<String>,
    #[serde(default)]
    pub scan_warning: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CalendarConversationVersion {
    pub conversation_id: String,
    pub title: String,
    pub last_activity_at: Option<DateTime<Utc>>,
    pub analysis_state: CalendarAnalysisState,
    pub snapshot_count: usize,
    pub is_latest: bool,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CalendarQuery {
    pub start_date: String,
    pub end_date_exclusive: String,
    pub time_zone: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CalendarIndexScanStatus {
    Ready,
    Partial,
    Failed,
}

impl CalendarIndexScanStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Ready => "ready",
            Self::Partial => "partial",
            Self::Failed => "failed",
        }
    }
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum CodexSessionKind {
    Primary,
    Internal,
    #[default]
    Unknown,
}

impl CodexSessionKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Primary => "primary",
            Self::Internal => "internal",
            Self::Unknown => "unknown",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CodexSessionIndexRecord {
    pub session_id: String,
    pub canonical_path: String,
    pub title: String,
    pub last_activity_at: Option<DateTime<Utc>>,
    pub last_completed_turn_at: Option<DateTime<Utc>>,
    pub last_message_id: Option<String>,
    pub source_state: CalendarSourceState,
    pub source_file_size: u64,
    pub source_file_mtime_ns: i64,
    pub scan_status: CalendarIndexScanStatus,
    pub session_id_inferred: bool,
    #[serde(default)]
    pub session_kind: CodexSessionKind,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct SourceSpan {
    pub message_id: String,
    pub start_utf16: usize,
    pub end_utf16: usize,
    pub exact_quote: String,
    pub sha256: String,
    #[serde(default)]
    pub model_saw_redacted: bool,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Provenance {
    Model,
    DeterministicFallback,
    User,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SemanticUnit {
    pub id: String,
    pub turn_id: String,
    pub speaker: Speaker,
    pub label: String,
    pub acts: Vec<String>,
    pub importance: f32,
    pub provenance: Provenance,
    pub source_spans: Vec<SourceSpan>,
    pub primary: bool,
    pub operation_only: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Relation {
    pub id: String,
    pub source: String,
    pub target: String,
    pub kind: String,
    pub label: String,
    pub confidence: f32,
    pub evidence: Vec<SourceSpan>,
    #[serde(default)]
    pub user_created: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Mode {
    pub id: String,
    pub kind: String,
    pub label: String,
    pub color: String,
    pub confidence: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ModeMembership {
    pub id: String,
    pub mode_id: String,
    pub unit_id: String,
    pub confidence: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IssueSeverity {
    Warning,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ValidationIssue {
    pub stage: String,
    pub item_id: Option<String>,
    pub severity: IssueSeverity,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisSnapshot {
    pub id: String,
    pub run_id: String,
    pub conversation_id: String,
    #[serde(default)]
    pub provider: AnalysisProviderKind,
    #[serde(default)]
    pub provider_version: Option<String>,
    #[serde(default)]
    pub credential_mode: Option<String>,
    pub model_id: String,
    pub prompt_version: String,
    pub schema_version: String,
    pub status: AnalysisState,
    pub semantic_units: Vec<SemanticUnit>,
    pub relations: Vec<Relation>,
    pub modes: Vec<Mode>,
    pub memberships: Vec<ModeMembership>,
    pub validation_issues: Vec<ValidationIssue>,
    /// Exact strict-JSON values returned by each model stage, before local filtering.
    #[serde(default)]
    pub raw_model_output: Value,
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AnalysisState {
    Parsing,
    PrivacyReview,
    Queued,
    Segmenting,
    Linking,
    Modes,
    Validating,
    Ready,
    Partial,
    Failed,
    Cancelled,
}

impl AnalysisState {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Parsing => "parsing",
            Self::PrivacyReview => "privacy_review",
            Self::Queued => "queued",
            Self::Segmenting => "segmenting",
            Self::Linking => "linking",
            Self::Modes => "modes",
            Self::Validating => "validating",
            Self::Ready => "ready",
            Self::Partial => "partial",
            Self::Failed => "failed",
            Self::Cancelled => "cancelled",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisProgress {
    pub run_id: String,
    pub conversation_id: String,
    pub stage: AnalysisState,
    pub progress: f32,
    pub completed: usize,
    pub total: usize,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CorrectionCommand {
    pub kind: String,
    pub target_id: String,
    pub after: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CorrectionEvent {
    pub id: String,
    pub snapshot_id: String,
    pub kind: String,
    pub target_id: String,
    pub before: Option<Value>,
    pub after: Value,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NodeLayout {
    pub unit_id: String,
    pub x: f64,
    pub y: f64,
    #[serde(default)]
    pub pinned: bool,
    #[serde(default)]
    pub collapsed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ViewportState {
    pub x: f64,
    pub y: f64,
    pub zoom: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct LayoutState {
    pub nodes: Vec<NodeLayout>,
    pub viewport: ViewportState,
    #[serde(default = "default_true")]
    pub show_mode_islands: bool,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SnapshotBundle {
    pub conversation: ConversationSummary,
    pub base: AnalysisSnapshot,
    pub effective: AnalysisSnapshot,
    pub corrections: Vec<CorrectionEvent>,
    pub layout: Option<LayoutState>,
    pub messages: Vec<SourceMessage>,
    pub turns: Vec<VisibleTurn>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisStart {
    pub run_id: String,
    pub conversation_id: String,
    pub state: AnalysisState,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ApiKeyStatus {
    pub ok: bool,
    pub configured: bool,
    pub valid: Option<bool>,
    pub model: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct LayoutItemInput {
    pub x: f64,
    pub y: f64,
    #[serde(default)]
    pub pinned: bool,
}

File: src/domain.ts
import type { XYPosition } from "@xyflow/react";

export type Speaker = "user" | "assistant" | "system";
export type NodeKind = "anchor" | "card" | "operation" | "unresolved";

export type DialogueAct =
  | "提问"
  | "请求"
  | "任务"
  | "建议"
  | "陈述"
  | "回答"
  | "解释"
  | "论证"
  | "举例"
  | "评价"
  | "同意"
  | "质疑"
  | "纠正"
  | "承诺"
  | "反馈"
  | "话语管理"
  | "约束"
  | "假设检验"
  | "证据"
  | "区分"
  | "归类"
  | "反例"
  | "限定"
  | "撤回"
  | "其他";

export type RelationType =
  | "回应"
  | "支持"
  | "理由"
  | "举例"
  | "条件"
  | "对比"
  | "质疑"
  | "反证"
  | "收窄"
  | "修正"
  | "重新归类"
  | "撤回"
  | "重新打开"
  | "中断后续答"
  | "导致"
  | "未解决";

export type ModeKind =
  | "目标定位"
  | "探索"
  | "方案形成"
  | "证据核验"
  | "质疑校正"
  | "决定"
  | "执行"
  | "协调"
  | "元对话"
  | "未分类";

export interface SourceSpan {
  messageId: string;
  start: number;
  end: number;
  exactQuote: string;
  sha256?: string;
  redacted?: boolean;
}

export interface EvidencePair {
  title?: string;
  user?: SourceSpan;
  assistant?: SourceSpan;
  context?: string;
}

export interface SemanticUnit {
  id: string;
  turnId: string;
  turnOrdinal: number;
  segmentOrdinal?: number;
  speaker: Speaker;
  kind: NodeKind;
  label: string;
  fullText: string;
  acts: DialogueAct[];
  importance: number;
  provenance: "model" | "fallback" | "user" | "fixture";
  sourceSpans: SourceSpan[];
  modeIds: string[];
  secondary?: boolean;
  state?: "active" | "downgraded" | "withdrawn" | "open";
}

export interface AtlasRelation {
  id: string;
  source: string;
  target: string;
  type: RelationType;
  label?: string;
  confidence: number;
  evidence: EvidencePair;
  provenance: "model" | "user" | "fixture";
}

export interface ModeDefinition {
  id: string;
  kind: ModeKind;
  label: string;
  color: string;
  confidence: number;
  inferred: boolean;
}

export interface LayoutItem extends XYPosition {
  pinned?: boolean;
}

export interface ConversationMeta {
  id: string;
  title: string;
  turns: number;
  totalUnits: number;
  expandedUnits: number;
  hiddenUnits: number;
  sourceKind: "codex_jsonl" | "paste" | "demo";
  sourcePath?: string;
}

export interface ValidationIssue {
  stage: string;
  itemId?: string;
  severity: "warning" | "error";
  message: string;
}

export interface AtlasSnapshot {
  id: string;
  conversation: ConversationMeta;
  units: SemanticUnit[];
  relations: AtlasRelation[];
  modes: ModeDefinition[];
  layout: Record<string, LayoutItem>;
  viewport?: { x: number; y: number; zoom: number };
  showModeIslands?: boolean;
  modelId: string;
  provider?: AnalysisProvider | "fixture";
  providerVersion?: string;
  credentialMode?: string;
  promptVersion: string;
  schemaVersion: string;
  createdAt: string;
  status?: "ready" | "partial";
  validationIssues?: ValidationIssue[];
  inputTokens?: number;
  outputTokens?: number;
  rawModelOutput?: unknown;
  sourceMessages?: Array<{
    id: string;
    speaker: "user" | "assistant";
    phase?: string;
    text: string;
  }>;
  visibleTurns?: Array<{
    id: string;
    ordinal: number;
    speaker: "user" | "assistant";
    operationOnly: boolean;
    messageIds: string[];
  }>;
}

export interface PreviewMessage {
  id: string;
  speaker: Speaker;
  phase?: "commentary" | "final";
  turnOrdinal: number;
  text: string;
  operationOnly?: boolean;
  redactions?: Array<{ start: number; end: number; replacement: string; kind: string }>;
}

export interface ImportPreview {
  id: string;
  title: string;
  sourceKind: "codex_jsonl" | "paste";
  sourcePath?: string;
  messages: PreviewMessage[];
  characterCount: number;
  warnings: string[];
  sourceFormat?: SourceFormat;
  externalSessionId?: string;
  firstVisibleAt?: string;
  lastActivityAt?: string;
  lastCompletedTurnAt?: string;
  timeCoverage?: TimeCoverage;
  sourceStillWriting?: boolean;
}

export type TimeStatus = "valid" | "missing" | "invalid";
export type TimeCoverage = "complete" | "partial" | "none";
export type SourceFormat = "raw_rollout" | "visible_export" | "paste" | "legacy_unknown";

export type CalendarCompletionState = "completed" | "in_progress_or_unknown" | "undated";
export type CalendarSourceState = "active" | "archived" | "missing" | "import_only";
export type CalendarImportState = "not_imported" | "imported_current" | "source_updated";
export type CalendarAnalysisState = "none" | "ready" | "partial" | "failed";

export interface CalendarEntry {
  id: string;
  externalSessionId?: string;
  title: string;
  lastActivityAt?: string;
  lastCompletedTurnAt?: string;
  completionState: CalendarCompletionState;
  sourceState: CalendarSourceState;
  importState: CalendarImportState;
  analysisState: CalendarAnalysisState;
  importedVersionCount: number;
  snapshotCount: number;
  latestConversationId?: string;
  turnCount?: number;
  activeDayCount?: number;
  timeCoverage?: TimeCoverage;
  sourcePath?: string;
  scanWarning?: string;
}

export interface CalendarConversationVersion {
  conversationId: string;
  title: string;
  lastActivityAt?: string;
  analysisState: CalendarAnalysisState;
  snapshotCount: number;
  isLatest: boolean;
  createdAt: string;
}

export interface CalendarQuery {
  startDate: string;
  endDateExclusive: string;
  timeZone: "Asia/Tokyo";
}

export interface CodexIndexProgress {
  stage: "idle" | "discovering" | "scanning" | "committing" | "ready" | "cancelled" | "failed";
  completed: number;
  total: number;
  message: string;
  visibleSessions?: number;
  skippedSessions?: number;
}

export interface CodexIndexStatus extends CodexIndexProgress {
  running: boolean;
  lastCompletedAt?: string;
}

export interface ImportPreviewProgress {
  previewId: string;
  completedBytes: number;
  totalBytes: number;
  message: string;
}

export interface ImportPreviewReady {
  previewId: string;
  preview?: ImportPreview;
  error?: string;
}

export type AnalysisStage =
  | "idle"
  | "parsing"
  | "privacy_review"
  | "segmenting"
  | "linking"
  | "modes"
  | "validating"
  | "ready"
  | "partial"
  | "failed"
  | "cancelled";

export interface AnalysisProgress {
  runId: string;
  conversationId: string;
  stage: AnalysisStage;
  completed: number;
  total: number;
  message: string;
}

export type AnalysisTaskStatus = "running" | "stopping" | "ready" | "partial" | "failed" | "cancelled";

export interface AnalysisTask {
  runId: string;
  conversationId: string;
  originEntryId?: string;
  title?: string;
  status: AnalysisTaskStatus;
  progress: AnalysisProgress;
  startedAt: string;
  updatedAt: string;
}

export type AnalysisProvider = "codex_cli" | "openai_api";

export type PlatformKind = "macos" | "windows" | "other";

export type CredentialStoreKind =
  | "macos_keychain"
  | "windows_credential_manager"
  | "system_keyring";

export interface PlatformCapabilities {
  platform: PlatformKind;
  availableProviders: AnalysisProvider[];
  credentialStore: CredentialStoreKind;
}

export interface AnalysisSettings {
  provider: AnalysisProvider;
  defaultOpenaiModel: string;
  codexCliModel: string;
  capabilities: PlatformCapabilities;
}

export interface AnalysisProviderStatus {
  provider: AnalysisProvider;
  ok: boolean;
  configured: boolean;
  available: boolean;
  authenticated: boolean;
  model: string;
  version?: string;
  message: string;
}

export const DEFAULT_ANALYSIS_SETTINGS: AnalysisSettings = {
  provider: "openai_api",
  defaultOpenaiModel: "gpt-5-mini",
  codexCliModel: "gpt-5.6-luna",
  // Loading starts fail-closed: the desktop backend replaces this with its
  // authoritative platform capability set before exposing any extra provider.
  capabilities: {
    platform: "other",
    availableProviders: ["openai_api"],
    credentialStore: "system_keyring",
  },
};

export type Selection =
  | { kind: "node"; id: string }
  | { kind: "edge"; id: string }
  | null;

export type CorrectionCommand =
  | { kind: "update_unit"; unitId: string; label: string; acts: DialogueAct[]; modeIds: string[] }
  | { kind: "update_relation"; relationId: string; type: RelationType; label?: string }
  | { kind: "add_relation"; relation: AtlasRelation }
  | { kind: "delete_relation"; relationId: string }
  | { kind: "update_mode"; modeId: string; label: string }
  | { kind: "move_node"; unitId: string; position: XYPosition; pinned: boolean };

export const DIALOGUE_ACTS: DialogueAct[] = [
  "提问",
  "请求",
  "任务",
  "建议",
  "陈述",
  "回答",
  "解释",
  "论证",
  "举例",
  "评价",
  "同意",
  "质疑",
  "纠正",
  "承诺",
  "反馈",
  "话语管理",
  "约束",
  "假设检验",
  "证据",
  "区分",
  "归类",
  "反例",
  "限定",
  "撤回",
  "其他",
];

export const RELATION_TYPES: RelationType[] = [
  "回应",
  "支持",
  "理由",
  "举例",
  "条件",
  "对比",
  "质疑",
  "反证",
  "收窄",
  "修正",
  "重新归类",
  "撤回",
  "重新打开",
  "中断后续答",
  "导致",
  "未解决",
];

export const RELATION_COLORS: Record<RelationType, string> = {
  回应: "#1773eb",
  支持: "#168f78",
  理由: "#0a9a93",
  举例: "#3191b7",
  条件: "#4d932f",
  对比: "#7e52c8",
  质疑: "#7c4fd0",
  反证: "#ef3e3e",
  收窄: "#774dd0",
  修正: "#774dd0",
  重新归类: "#7546d6",
  撤回: "#c9202b",
  重新打开: "#f13f3f",
  中断后续答: "#1773eb",
  导致: "#118986",
  未解决: "#667085",
};

export const MAX_VISIBLE_GRAPH_NODES = 120;
